### PR TITLE
GH-44952: [C++][Python] Add Hyperbolic Trig functions

### PR DESCRIPTION
**Pull request by khwilson**
### Rationale for this change

Hyperbolic trigonometric functions are a common transformation for dealing with skewed data. And they are built into the core C++ libraries so require minimal change.

### What changes are included in this PR?

Adding `a?(sin|cos|tan)h` to the base C++ library and substrait and tests for these functions in pyarrow.

### Are these changes tested?

Yes, in the same style as the trigonometric functions.

### Are there any user-facing changes?

Yes. Additional compute functions are added.
* GitHub Issue: #44952

### CODE REVIEW

**Comment by khwilson** [comment]
Hi @amoeba just wanted to check in if you had a chance to review this commit. The test failure is very strange (on an encryption test?!), but I can look into it!

**Comment by adamreeve** [comment]
> The test failure is very strange (on an encryption test?!), but I can look into it!

This test is known to be flaky (#43057), the failure is unrelated to the changes in this PR.

**Comment by amoeba** [comment]
Hi @khwilson, thanks for the ping. I'll take a look in the next couple of days.

File: cpp/src/arrow/engine/substrait/extension_set.cc
Comment (amoeba) [inline cpp/src/arrow/engine/substrait/extension_set.cc:1211]: Thanks for catching that we can add these now, I probably would've missed it in review. No changes needed here.

**Review by amoeba — CHANGES_REQUESTED** [review]
Thanks for contributing this. I left some notes inline.

Beyond that, can you please update the [Python API docs](https://arrow.apache.org/docs/python/api/compute.html#trigonometric-functions) to list these new functions? 

(See https://github.com/apache/arrow/commit/dfb0928e91c0d3bd89cb0497a3948ed8fea7fc78#diff-ce5b94577014735990903d3d03bd4ea4b8c8e6d32f5227592e60b7dd6a912d59R342 for an example.)

**Comment by amoeba** [comment]
@felipecrv do you have any time to give this a second review?

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc
Comment (felipecrv) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:335]: No, it's actually a domain error:

```
Cell In[24], line 1
----> 1 math.atanh(-1.0)

ValueError: math domain error

In [25]: math.atanh(1.0)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[25], line 1
----> 1 math.atanh(1.0)

ValueError: math domain error
```

Comment (felipecrv) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:335]:
```suggestion
    if (ARROW_PREDICT_FALSE((val <= -1.0 || val => 1.0))) {
```

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc
Comment (felipecrv) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:346]:
```suggestion
    if (ARROW_PREDICT_FALSE((val <= -1.0 || val => 1.0))) {
```

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc
Comment (felipecrv) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:1302]:
```suggestion
                            ("NaN is returned for input values < 1.0;\n"
```

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc
Comment (felipecrv) [inline cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:1323]:
```suggestion
const FunctionDoc atanh_doc{"Compute the inverse hyperbolic tangent",
                            ("NaN is returned for input values outside of the (-1.0, 1.0) range;\n"
                             "to raise an error instead, see \"atanh_checked\"."),
                            {"x"}};
```

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc
Comment (felipecrv) [inline cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:1328]:
```suggestion
const FunctionDoc atanh_checked_doc{"Compute the inverse hyperbolic tangent",
                                    ("Input values outside the (-1.0, 1.0) range raise an error;\n"
                                     "to return NaN instead, see \"atanh\"."),
                                    {"x"}};
```

File: cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc
Comment (felipecrv) [inline cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc:2546]: add -1 and 1 as expecting NaN as well.

**Review by felipecrv — CHANGES_REQUESTED** [review]
This is very well done. I found a simple range check error and suggested documentation improvements.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc
Comment (khwilson) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:335]: I actually went back and forth on this. In *python* they raise a domain error. In the C++ stdlib, they return ± infinity. So the code as is returns ± infinity and an error is raised (which I realize I didn't clear....). See [here](https://en.cppreference.com/w/cpp/numeric/math/atanh).

Which library this PR should be consistent with

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc
Comment (khwilson) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:346]: Same comment as the previous one

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc
Comment (khwilson) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:1302]: I was following the convention of the other inverse trig functions. Should we switch their docs at the same time to specify their ranges?

**Comment by khwilson** [comment]
Thanks for the reviews. I can:
* Fix up the Python docs;
* Update the C++ docs for all the trig functions to reflect their domains instead of just "invalid inputs."
* Sort the code as suggested.

Remaining question for reviewers. In C++, `atanh(1) = + infinity` (and sets a divide by zero error). In Python `atanh(1)` is a domain error.

Which behaviour should this library (which spans both C++ and Python) utilize?

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc
Comment (felipecrv) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:335]: C++ docs also say `If the argument is ±1, a pole error occurs.` and `if the argument is ±1, ±∞ is returned and [FE_DIVBYZERO] is raised.

They have different names for the errors (range/pole/FE_DIVBYZERO) but that doesn't change the fact that Mathematically, `atanh` isn't defined at the -1.0 and 1.0 points.

The function **converges to** $\pm\infty$ as $x$ approaches the extremes of the range which makes returning `-Inf/+Inf` more reasonable than returning `NaN` at these extremes, but it's not defined at these points per se.

$$\lim_{x \to 1.0}{tanh^{-1}(x)} = \infty$$
$$\lim_{x \to -1.0}{tanh^{-1}(x)} = -\infty$$

So I think you can drop my suggestion on the unchecked version (this one) and let the implementation return `-Inf/+Inf` instead of `NaN` at these extremes, but the `atanh_checked` variation should fail at `-1.0/+1.0` instead of returning the infinite values.

Docs should be updated accordingly to describe this behavior.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc
Comment (felipecrv) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:346]: The checked version should reject -1.0 and 1.0. Justification in the previous thread.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc
Comment (felipecrv) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:1302]: No need. Just make the ones you are adding more informative.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc
Comment (felipecrv) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:1323]: Updated the suggestion to mention -Inf/Inf

**Comment by felipecrv** [comment]
> Remaining question for reviewers. In C++, `atanh(1) = + infinity` (and sets a divide by zero error). In Python `atanh(1)` is a domain error.
> 
> Which behaviour should this library (which spans both C++ and Python) utilize?

Answered in the respective threads. Summary: -Inf/Inf on the unchecked version, range error on the checked version.

**Comment by khwilson** [comment]
Makes sense. Will implement.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc
Comment (ianmcook) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:335]: FWIW, a common approach when implementing Acero functions in cases like this has been to add options to allow the user can pick what behavior they want. It might be worth doing that here if it doesn't add excessive complexity or inefficiency.

Comment (felipecrv) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:335]: The logarithms definition of `atanh` makes the issue more apparent:

$$tanh^{-1}(x) = \frac{1}{2}\ln\frac{1 + x}{1 - x}$$

If x is 1, you get a division by 0. If x is -1 you get $\ln(0)$ which is undefined (converges to `-Inf`, but not defined).

Comment (khwilson) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:335]: Oh totally! I honestly think both implementations can be rigorously justified and defended.

For instance, I could *define* atanh as the inverse of tanh. And I could *define* f(oo) = lim_{x -> oo} f(x). And then atanh(1) is clearly oo.

Just a matter of which convention the library wants to follow 😀

File: cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc
Comment (khwilson) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc:2414]: Done.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc
Comment (khwilson) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:1253]: Whoops.

File: cpp/src/arrow/compute/api_scalar.cc
Comment (khwilson) [reply cpp/src/arrow/compute/api_scalar.cc:753]: Done!

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc
Comment (khwilson) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:335]: Notably, C++ defines [tanh(oo) = 1](https://en.cppreference.com/w/cpp/numeric/math/tanh)

Comment (amoeba) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:335]: Thanks for digging in all. I think the direction this conversion went makes sense (unchecked version returns -Inf/Inf, checked version raises). I was checking these in R and SQLite and it looks like Python is sort of the odd man out here.

Comment (khwilson) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:335]: Sounds good. Change made.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc
Comment (khwilson) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:1302]: Done.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc
Comment (khwilson) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc]: Chose slightly different wording, but same spriti.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc
Comment (khwilson) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc:2545]: That happened a few lines below. But now that they behave differently I've split out the test on lines 2546 and 2553.

**Comment by khwilson** [comment]
OK, I believe all the comments were addressed. Now to wait for tests and likely rebase :-)

**Review by felipecrv — APPROVED** [review]
I can merge soon if no one has more changes to request.

**Review by felipecrv — CHANGES_REQUESTED** [review]
Please fix the compilation errors. >= and not =>

**Comment by khwilson** [comment]
Well, that's embarrassing. Pushed the fix.

**Comment by khwilson** [comment]
Also rebased onto main as a bunch of tests seem to be failing for reasons I'm guessing are due to being slightly out of date :-)

**Comment by khwilson** [comment]
Huh, looks like azurite is [currently 404-ing](https://www.npmjs.com/package/azurite), which is why some of the tests are failing.

**Comment by amoeba** [comment]
I'm keeping an eye on them and will re-run the unrelated failures as needed.

**Comment by amoeba** [comment]
@khwilson: This looks like a real failure: https://github.com/apache/arrow/actions/runs/12224231650/job/34096726088?pr=44630#step:6:3100. I think that can be fixed by adding a `\n`.

**Comment by khwilson** [comment]
Added. Ran the main tests locally, but didn't run the doc tests it seems.

**Review by amoeba — APPROVED** [review]
LGTM and CI is green. I'll let @felipecrv merge.

**Pull request merged**
